File: codex-rs/core/src/tools/parallel.rs
use std::sync::Arc;
use std::time::Instant;

use tokio::sync::RwLock;
use tokio::task::JoinError;
use tokio_util::either::Either;
use tokio_util::sync::CancellationToken;
use tokio_util::task::AbortOnDropHandle;
use tracing::Instrument;
use tracing::instrument;
use tracing::trace_span;

use crate::function_tool::FunctionCallError;
use crate::session::session::Session;
use crate::session::turn_context::TurnContext;
use crate::tools::context::AbortedToolOutput;
use crate::tools::context::SharedTurnDiffTracker;
use crate::tools::context::ToolPayload;
use crate::tools::registry::AnyToolResult;
use crate::tools::registry::ToolArgumentDiffConsumer;
use crate::tools::router::ToolCall;
use crate::tools::router::ToolCallSource;
use crate::tools::router::ToolRouter;
use codex_protocol::error::CodexErr;
use codex_protocol::models::ResponseInputItem;
use codex_tools::ToolSpec;

#[derive(Clone)]
pub(crate) struct ToolCallRuntime {
    router: Arc<ToolRouter>,
    session: Arc<Session>,
    turn_context: Arc<TurnContext>,
    tracker: SharedTurnDiffTracker,
    parallel_execution: Arc<RwLock<()>>,
}

impl ToolCallRuntime {
    pub(crate) fn new(
        router: Arc<ToolRouter>,
        session: Arc<Session>,
        turn_context: Arc<TurnContext>,
        tracker: SharedTurnDiffTracker,
    ) -> Self {
        Self {
            router,
            session,
            turn_context,
            tracker,
            parallel_execution: Arc::new(RwLock::new(())),
        }
    }

    pub(crate) fn find_spec(&self, tool_name: &codex_tools::ToolName) -> Option<ToolSpec> {
        self.router.find_spec(tool_name)
    }

    pub(crate) fn create_diff_consumer(
        &self,
        tool_name: &codex_tools::ToolName,
    ) -> Option<Box<dyn ToolArgumentDiffConsumer>> {
        self.router.create_diff_consumer(tool_name)
    }

    #[instrument(level = "trace", skip_all)]
    pub(crate) fn handle_tool_call(
        self,
        call: ToolCall,
        cancellation_token: CancellationToken,
    ) -> impl std::future::Future<Output = Result<ResponseInputItem, CodexErr>> {
        let error_call = call.clone();
        let future =
            self.handle_tool_call_with_source(call, ToolCallSource::Direct, cancellation_token);
        async move {
            match future.await {
                Ok(response) => Ok(response.into_response()),
                Err(FunctionCallError::Fatal(message)) => Err(CodexErr::Fatal(message)),
                Err(other) => Ok(Self::failure_response(error_call, other)),
            }
        }
        .in_current_span()
    }

    #[instrument(level = "trace", skip_all)]
    pub(crate) fn handle_tool_call_with_source(
        self,
        call: ToolCall,
        source: ToolCallSource,
        cancellation_token: CancellationToken,
    ) -> impl std::future::Future<Output = Result<AnyToolResult, FunctionCallError>> {
        let supports_parallel = self.router.tool_supports_parallel(&call);
        let router = Arc::clone(&self.router);
        let session = Arc::clone(&self.session);
        let turn = Arc::clone(&self.turn_context);
        let tracker = Arc::clone(&self.tracker);
        let lock = Arc::clone(&self.parallel_execution);
        let invocation_cancellation_token = cancellation_token.clone();
        let started = Instant::now();
        let display_name = call.tool_name.display();
        let join_error_call = call.clone();

        let dispatch_span = trace_span!(
            "dispatch_tool_call_with_code_mode_result",
            otel.name = display_name.as_str(),
            tool_name = display_name.as_str(),
            call_id = call.call_id.as_str(),
            aborted = false,
        );

        let handle: AbortOnDropHandle<Result<AnyToolResult, FunctionCallError>> =
            AbortOnDropHandle::new(tokio::spawn(async move {
                tokio::select! {
                    _ = cancellation_token.cancelled() => {
                        let secs = started.elapsed().as_secs_f32().max(0.1);
                        dispatch_span.record("aborted", true);
                        Ok(Self::aborted_response(&call, secs))
                    },
                    res = async {
                        let _guard = if supports_parallel {
                            Either::Left(lock.read().await)
                        } else {
                            Either::Right(lock.write().await)
                        };

                        router
                            .dispatch_tool_call_with_code_mode_result(
                                session,
                                turn,
                                invocation_cancellation_token,
                                tracker,
                                call.clone(),
                                source,
                            )
                            .instrument(dispatch_span.clone())
                            .await
                    } => res,
                }
            }));

        async move {
            handle.await.map_err(|err| {
                tool_task_join_error_to_function_call_error(
                    err,
                    &join_error_call,
                    started.elapsed(),
                )
            })?
        }
        .in_current_span()
    }
}

fn tool_task_join_error_to_function_call_error(
    err: JoinError,
    call: &ToolCall,
    elapsed: std::time::Duration,
) -> FunctionCallError {
    if err.is_cancelled() {
        let secs = elapsed.as_secs_f32().max(0.1);
        tracing::warn!(
            tool_name = %call.tool_name,
            call_id = call.call_id,
            elapsed_seconds = secs,
            "tool task was cancelled before delivering a response"
        );
        return FunctionCallError::RespondToModel(format!(
            "tool execution interrupted after {secs:.1}s"
        ));
    }

    FunctionCallError::Fatal(format!("tool task failed to receive: {err:?}"))
}

impl ToolCallRuntime {
    fn failure_response(call: ToolCall, err: FunctionCallError) -> ResponseInputItem {
        let message = err.to_string();
        match call.payload {
            ToolPayload::ToolSearch { .. } => ResponseInputItem::ToolSearchOutput {
                call_id: call.call_id,
                status: "completed".to_string(),
                execution: "client".to_string(),
                tools: Vec::new(),
            },
            ToolPayload::Custom { .. } => ResponseInputItem::CustomToolCallOutput {
                call_id: call.call_id,
                name: None,
                output: codex_protocol::models::FunctionCallOutputPayload {
                    body: codex_protocol::models::FunctionCallOutputBody::Text(message),
                    success: Some(false),
                },
            },
            _ => ResponseInputItem::FunctionCallOutput {
                call_id: call.call_id,
                output: codex_protocol::models::FunctionCallOutputPayload {
                    body: codex_protocol::models::FunctionCallOutputBody::Text(message),
                    success: Some(false),
                },
            },
        }
    }

    fn aborted_response(call: &ToolCall, secs: f32) -> AnyToolResult {
        AnyToolResult {
            call_id: call.call_id.clone(),
            payload: call.payload.clone(),
            result: Box::new(AbortedToolOutput {
                message: Self::abort_message(call, secs),
            }),
            post_tool_use_payload: None,
        }
    }

    fn abort_message(call: &ToolCall, secs: f32) -> String {
        if call.tool_name.namespace.is_none()
            && matches!(
                call.tool_name.name.as_str(),
                "shell" | "container.exec" | "local_shell" | "shell_command" | "unified_exec"
            )
        {
            format!("Wall time: {secs:.1} seconds\naborted by user")
        } else {
            format!("aborted by user after {secs:.1}s")
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tools::context::ToolPayload;
    use pretty_assertions::assert_eq;
    use serde_json::json;

    fn test_tool_call() -> ToolCall {
        ToolCall {
            tool_name: "spawn_agents_on_csv".into(),
            call_id: "call-1".to_string(),
            payload: ToolPayload::Function {
                arguments: json!({"csv_path":"in.csv","instruction":"test"}).to_string(),
            },
        }
    }

    #[tokio::test]
    async fn cancelled_tool_task_join_error_becomes_model_response() {
        let handle = tokio::spawn(async {
            tokio::time::sleep(std::time::Duration::from_secs(60)).await;
        });
        handle.abort();
        let err = handle.await.expect_err("task should be cancelled");

        let mapped = tool_task_join_error_to_function_call_error(
            err,
            &test_tool_call(),
            std::time::Duration::from_millis(200),
        );

        assert_eq!(
            mapped,
            FunctionCallError::RespondToModel("tool execution interrupted after 0.2s".to_string())
        );
    }

    #[tokio::test]
    async fn panicked_tool_task_join_error_stays_fatal() {
        let handle = tokio::spawn(async {
            panic!("boom");
        });
        let err = handle.await.expect_err("task should panic");

        let mapped = tool_task_join_error_to_function_call_error(
            err,
            &test_tool_call(),
            std::time::Duration::from_millis(200),
        );

        let FunctionCallError::Fatal(message) = mapped else {
            panic!("panic join errors should remain fatal");
        };
        assert!(message.contains("tool task failed to receive"));
    }
}

File: src/nativeMain/kotlin/ai/solace/coder/core/tools/Parallel.kt
// port-lint: source core/src/tools/parallel.rs
package ai.solace.coder.core.tools

import ai.solace.coder.core.error.CodexError
import ai.solace.coder.core.session.Session
import ai.solace.coder.core.session.TurnContext
import ai.solace.coder.protocol.FunctionCallOutputPayload
import ai.solace.coder.protocol.ResponseInputItem
import kotlinx.coroutines.Job
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.selects.select
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.time.TimeSource

/**
 * A ReadWriteMutex that allows multiple readers or a single writer.
 * Matches tokio::sync::RwLock semantics used in Rust.
 */
class ReadWriteMutex {
    private val writeMutex = Mutex()
    private val readMutex = Mutex()
    private var readers = 0

    suspend fun <T> read(action: suspend () -> T): T {
        readMutex.lock()
        try {
            if (readers == 0) {
                writeMutex.lock()
            }
            readers++
        } finally {
            readMutex.unlock()
        }

        try {
            return action()
        } finally {
            readMutex.lock()
            try {
                readers--
                if (readers == 0) {
                    writeMutex.unlock()
                }
            } finally {
                readMutex.unlock()
            }
        }
    }

    suspend fun <T> write(action: suspend () -> T): T {
        writeMutex.withLock {
            return action()
        }
    }
}

class ToolCallRuntime(
    private val router: ToolRouter,
    private val session: Session,
    private val turnContext: TurnContext,
    private val tracker: SharedTurnDiffTracker
) {
    private val parallelExecution = ReadWriteMutex()

    suspend fun handleToolCall(
        call: ToolCall,
        cancellationJob: Job?
    ): Result<ResponseInputItem> {
        val supportsParallel = router.toolSupportsParallel(call.toolName)
        val started = TimeSource.Monotonic.markNow()

        return try {
            coroutineScope {
                var result: Result<ResponseInputItem>? = null
                
                val job = launch {
                    if (supportsParallel) {
                        parallelExecution.read {
                            result = router.dispatchToolCall(session, turnContext, tracker, call.copy())
                        }
                    } else {
                        parallelExecution.write {
                            result = router.dispatchToolCall(session, turnContext, tracker, call.copy())
                        }
                    }
                }

                if (cancellationJob != null) {
                    select {
                        cancellationJob.onJoin {
                            job.cancel()
                            val secs = started.elapsedNow().inWholeMilliseconds / 1000.0f
                            result = Result.success(abortedResponse(call, secs.coerceAtLeast(0.1f)))
                        }
                        job.join()
                    }
                } else {
                    job.join()
                }
                
                result ?: Result.failure(CodexError.Fatal("Tool execution failed to produce result"))
            }
        } catch (e: Exception) {
             Result.failure(CodexError.Fatal("tool task failed to receive: ${e.message}"))
        }
    }

    private fun abortedResponse(call: ToolCall, secs: Float): ResponseInputItem {
        return when (val payload = call.payload) {
            is ToolPayload.Custom -> ResponseInputItem.CustomToolCallOutput(
                callId = call.callId,
                output = abortMessage(call, secs)
            )
            is ToolPayload.Mcp -> ResponseInputItem.McpToolCallOutput(
                callId = call.callId,
                result = ai.solace.coder.protocol.Result(
                    value = null,
                    error = abortMessage(call, secs)
                )
            )
            else -> ResponseInputItem.FunctionCallOutput(
                callId = call.callId,
                output = FunctionCallOutputPayload(
                    content = abortMessage(call, secs),
                    success = false // Default
                )
            )
        }
    }

    private fun abortMessage(call: ToolCall, secs: Float): String {
        val formattedSecs = ((secs * 10).toInt() / 10.0).toString()
        return when (call.toolName) {
            "shell", "container.exec", "local_shell", "shell_command", "unified_exec" -> {
                "Wall time: $formattedSecs seconds\naborted by user"
            }
            else -> "aborted by user after ${formattedSecs}s"
        }
    }
}
